DOCUMENTATION.md

```diff
@@ -43,6 +45,153 @@ The overall territory is laid out here in this [C4](https://c4model.com) System
+
+### _Sink pipeline_
+
+'Sink' refers to the end of the processing pipeline that's not specific to a Store. It's primary role is to drive _handler_ invocations based incoming batches of events from the 'Source'. It encompasses: 
+
+   a. `Ingester`: responsible for limiting the maximum read-ahead per Tranche (one per Tranche)
+   b. `ProgressWriter`: used by the `Ingester` to hold, and periodically commit updates to the checkpoint for any progress that has been made (one per Tranche)
+   c. `Scheduler`: takes batches from the Ingester, buffering them. Continually feeds items to the `Dispatcher`, reporting latencies and outcomes (to `Stats`) and batch completion (to `Ingester`)
+   d. `Dispatcher`: handles keeping up to the desired amount of concurrent handlers in flight at any time
+   e. `Stats`: processes Handler invocation outcomes, maintaining statistics for periodic emission and/or forwarding to a metrics sinks such as `Prometheus` (can be customised to gather Domain-relevant statistics identifying the nature of the Reactions processing taking place in addition to generic invocation latency, stream counts and category distributions etc)
+   h. `Handler`s: caller-supplied function that's passed a stream identifier (a StreamName:- Category + Id pair) and the span of waiting events for that stream (typically a single event, but can grow in catch-up scenarios, retries, or multiple events being written in close succession)
+
+### Ordering
+
+Event ordering guarantees are a key consideration in running projections or reactions. The origin of the events will intrinsically dictate the upper bound of ordering guarantees possible. Some examples:
+- EventStore, SqlStreamStore: these establish a stable order as each event is written (within streams, but also across all streams). The '`$all`' stream and related APIs preserve this order
+- CosmosDb: the ChangeFeed guarantees to present Items from logical partitions in order of when they were touched (added/updated). While the relative order of writes across streams happens to be stable when there is one physical partition, that's an irrelevant implementation detail that's not useful for building anything other than a throwaway demo.
+- DynamoDb: While DynamoDb streams establishes and preserves the order of each change, that's subject to the 24h retention period limit. The DynamoStore.Indexer and reader all-but preserve that ordering (Note that the indexer does not currently ensure to preserve order _within_ any DDB Streams batch being indexed).
+- MessageDb: Appends establish a permanent position within the events for a given category (writes for a given category are serialized in order to guarantee this; there are explicitly no guarantees across categories). The Propulsion reader logic pulls categories independently (which should not matter due to the lack of correlation within the store).
```

Comment:
Small correction. There is a global order.

e.g. If I have an event with `global_position = 1500` in category `A`, and an event with `global_position = 1560` in category `B` I know that the event in `A` happened before the event in `B`. The reason for the category restriction is what happens when you're tailing the store under heavy load. Under those conditions a momentary gap can form in the global event sequence (`1,3` which later becomes `1,2,3`). If you happen to read a page while this gap exists you would process events 1 and 3, write your checkpoint as 3 and then never process event 2.

```suggestion
- MessageDb: Events technically have a global order but the store does not expose an `$all` stream equivalent. It does expose a way to read a given category in commit order. The Propulsion reader logic pulls categories independently, propulsion therefore provides no ordering guarantees across categories.
```

Reply:
I knew there was a global order (because you told me!). Was attempting to give only relevant information (because categories are read independently, there's no intrinsic global order that Propulsion could choose to preserve) without betraying my lack of complete knowledge. Thanks for the wording improvement - it's good to have the description be more complete (and vetted!).

Reply:
Added a commit after to try to cover your intro bit and what I was trying to convey. Its a mess so hopefully someobody will suggest something more succinct...